### ISSUE TITLE
Support configurability in base image sources

### ISSUE THREAD

**Issue by micahhausler** [kubernetes-sigs/kind#2030]
## What would you like to be added

The kind node base image today trusts multiple independent supply chains. Currently the [`Dockerfile`](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile) is based on Docker hub's Ubuntu container image ([`FROM ubuntu:20.10`](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile#L22)), uses [Canonical's packages](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile#L67-L72), and pulls binaries of [containerd](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile#L86), [crictl](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile#L97), and [cni](https://github.com/kubernetes-sigs/kind/blob/v0.10.0/images/base/Dockerfile#L103) from GitHub releases. 

### deb packages

I am not proposing any change today. Future changes could allow specifying an alternate package repository.

### base image
I'd like to make the `FROM` argument configurable

```Dockerfile
ARG BASE_IMAGE=ubuntu:20.10
FROM $BASE_IMAGE
```

### binaries
I'd like to make the the CNI, crictl, and containerd tarball URLs configurable. So something roughly like

```Dockerfile
ARG CONTAINERD_SHA256SUM=f42a4a943ef387914a66b8839ee0ab885513f2ea65cbf30b75bff4ccbb4ec72a
ARG CONTAINERD_URL=https://github.com/kind-ci/containerd-nightlies/releases/download/containerd-${CONTAINERD_VERSION#v}
```

## Why is this needed

I'd like to be able to build a node base image, but be able to specify sources that I trust and not need to maintain a fork of this repo. Note that this is only for the base image, not the final node image

All defaults would remain the same, but builders could specify alternate sources or base images with container build time arguments. Would the maintainers be open to a PR to the Dockerfile with these changes?

**Comment by BenTheElder** [comment]
TLDR: A resounding +1

---

The longer context ...

I see the rough proposal around binaries getting a little tricky for multi-arch (one env per hash per arch maybe? ugh), but I do think that we should be checking these AND that these should be configurable.

I think there might need to be a bit of "you ~break it~ replace the sources, you bought it" though 🙃 

As we develop kind we frequently pick up very recent releases of e.g. containerd to pickup bug-fixes and features that we leverage to keep kind functional (e.g. the ability to ignore huge pages in the CRI config, without which kind was broken on many hosts). Someone leveraging this proposal to override the binaries might have some friction with other aspects making assumptions about what versions of these tools we have, but as long as that's understood I think there's only a net-win here.

**Comment by BenTheElder** [comment]
I'd actually further add that while `kind build node-image` supports supplying an arbitrary base image, there's a tacit assumption that it's relatively equivalent in contents / functionality to the current default, but the flag exists both for quicker iteration of minor changes _and_ to enable this sort of "we must supply our own trusted image".

I know a few users at my employer would like better support for exactly this request 🙃 

Also: thank you for the lovely detailed issue! 😄

**Comment by micahhausler** [comment]
> I see the rough proposal around binaries getting a little tricky for multi-arch (one env per hash per arch maybe? ugh), but I do think that we should be checking these AND that these should be configurable.

A hash per $ARCH makes sense to me.

> I think there might need to be a bit of "you break it replace the sources, you bought it" though 🙃

Yep, that is my thinking as well. I will ensure that docs are updated appropriately
